core/src/main/java/com/linecorp/armeria/client/DefaultDnsQueryLifecycleObserver.java

```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright 2020 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package com.linecorp.armeria.client;
+
+import java.net.InetSocketAddress;
+import java.util.List;
+
+import io.micrometer.core.instrument.Counter;
+import io.micrometer.core.instrument.MeterRegistry;
+import io.netty.channel.ChannelFuture;
+import io.netty.handler.codec.dns.DnsQuestion;
+import io.netty.handler.codec.dns.DnsResponseCode;
+import io.netty.resolver.dns.DnsQueryLifecycleObserver;
+
+/**
+ * A {@link DnsQueryLifecycleObserver} that helps capture custom dns metrics.
+ */
+final class DefaultDnsQueryLifecycleObserver implements DnsQueryLifecycleObserver {
+
+    private final Counter success;
+    private final Counter failure;
+    private final Counter queryWritten;
+
+    private final DnsQuestion dnsQuestion;
+
+    /**
+     * Accepts meterRegistry.
+     * @param meterRegistry {@link MeterRegistry} MeterRegistry to capture metrics.
+     * @param question {@link DnsQuestion} DnsQuestion.
+     */
+    DefaultDnsQueryLifecycleObserver(MeterRegistry meterRegistry, DnsQuestion question) {
+        this.dnsQuestion = question;
+        success = meterRegistry.counter("dns.query", "result", "success");
+        failure = meterRegistry.counter("dns.query", "result", "failure");
+        queryWritten = meterRegistry.counter("dns.query", question.name(), "written");
```

Comment:
- We can add more metrics that are enumerated in #1887 such as protocol(UDP, TCP) and DNS error code and so on.
- Could you add test cases to verify this feature's correctness?

Reply:
- We could also add a tag for the type of the questions by using `DnsQuestion.type()`
- How about adding a tag for the domain names for all meters? I guess the cardinality will not be that bad.

Reply:
@ikhoon Good suggestions. I will implement them.
@trustin I was worried about the cardinality of adding domain names but it depends on the user of the client api. I think i can go ahead and add it. 
Is there a metrics consumer I can use to see if the metrics are showing up? I have a local unit test that I am using and i can verify in debugger that metrics are being populated. I guess I can extend the UT to add assertions etc.